### ISSUE TITLE
`git rev-parse --show-toplevel` expands to full path through `subst` drive location

### ISSUE THREAD

**Issue by alvinhochun** [git-for-windows/git#2705]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options
git version 2.27.0.windows.1
cpu: x86_64
built from commit: 907ab1011dce9112700498e034b974ba60f8b407
sizeof-long: 4
sizeof-size_t: 8
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

10.0.18362.900
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Performance Tweaks FSCache: Enabled
Use Credential Manager: Disabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

The Git repository in question is inside a `subst` drive (listed commands are for CMD):

1. Create a subst drive (`mkdir C:\subst_dir && subst Z: C:\subst_dir`)
2. Create a git repo in there (`mkdir Z:\repo && cd /d Z:\repo && git init && echo Hello > test.txt && git add test.txt && git commit -m first`)

This issue seems to be the apparent cause of https://github.com/microsoft/vscode/issues/100274

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash / CMD / vscode

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git rev-parse --show-toplevel
```
 - What did you expect to occur after running these commands?

It should print:
```
Z:/repo
```

 - What actually happened instead?

It prints:
```
C:/subst_dir/repo
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

n/a

**Comment by elieux** [comment]
This seems to have appeared fairly recently with some path canonization code if I remember correctly.  @dscho, do you remember if/where/when we talked about subst issues?

**Comment by dscho** [comment]
Yes, this was an intentional change: we want to resolve symbolic links efficiently, and as a consequence, `subst`ed drives are also resolved correctly.

My biggest question now is: why would this new behavior be considered a bug? A change in behavior I can understand. But a _bug_? What Git does is correct, after all, it canonicalizes the path.

**Comment by mfriedrich74** [comment]
Hi,

The root path presented to the user on command prompt by the OS/shell
should match the path provided to the user by git commands. That is the
usual user expectation.

Best regards, Mike

**Comment by alvinhochun** [comment]
It can certainly be considered a bug for some users because it breaks expectations of existing things that relies on the previous behaviour. Whether it is actually a bug is up for debate.

I just want the VSCode issue (https://github.com/microsoft/vscode/issues/100274) caused by this change to be fixed, and I am not familiar enough with either VSCode's Git extension or Git for Windows to know which one of the two should be patched. Currently I'm thinking of possible ways to fix it on VSCode's end. The thing is, canonicalizing paths is easy but the reverse is more difficult.

**Comment by mfriedrich74** [comment]
Canonicising isn't the same as resolving links and subst that the shell
isn't expanding.

Any tool or function calculating relative paths from two absolute paths
could fail here.

Canonicalisation should remove . and .., use of double slash, etc.
In my opinion it should not expand symbolic links.
If resolved to underlying file system file, both paths should be saved
internally, especially for NTFS junctions (where the user could not have
access to the source, but only to the path before resolving links - e.g.
for ntfs junctions on SMB shares).

**Comment by oscarfv** [comment]
After 20 years using `subst` this is the first time an application does not honor the user's wishes about having a drive letter for a directory. Right now `git` completely breaks my setup because it tells other applications that it is working on a directory different from what they expect.

On the linked vscode issue it is mentioned that this behavior started with `git 2.5`, so I downgraded to `2.24.1.windows.2` and rebooted, but that version exhibits the same problem.

**Comment by kcuzner** [comment]
I need to second that this is completely unintuitive behavior and does not honor the wishes of the user as to how their filesystem is presented. I've been scratching my head over this and other issues (probably stemming from this, my includeIf's in .gitconfig were malfunctioning if they referenced a subst drive) for over a year and just keeping myself on an old version of git until I was finally forced to upgrade and really dig into this today since I need the `git maintenance` command.

For me, this actually breaks interoperability of a tool I use with perforce (which I hate, for the record). I'm using git to manage perforce changelists and now the output of `git rev-parse --show-toplevel` can't be fed directly into the p4 CLI since it takes me at my word that the root of the depot is really `P:/` and not `D:/Perforce`.

**Comment by SiggyBar** [comment]
I use [persistent SUBSTs](https://github.com/ildar-shaimordanov/psubst) to allocate dedicated drive letters for some of my main Git projects, and I observe a similar non-expected behavior in that case.  Here I use a regular SUBST which also shows the problem:

```cmd
C:\>cmd.exe /c ver

Microsoft Windows [Version 10.0.19043.1415]

C:\>git --version
git version 2.34.1.windows.1

C:\>mkdir subst_dir

C:\>subst P: C:\subst_dir

C:\>P:

P:\>git init
Initialized empty Git repository in C:/subst_dir/.git/

P:\>git checkout P:\file
fatal: P:\file: 'P:\file' is outside repository at 'C:/subst_dir'

P:\>rmdir /s /q .git

P:\>mkdir sub_dir

P:\>cd sub_dir

P:\sub_dir>git init
Initialized empty Git repository in C:/subst_dir/sub_dir/.git/

P:\sub_dir>git checkout P:\sub_dir\file
error: pathspec 'P:\sub_dir\file' did not match any file(s) known to git
```

The odd thing is the difference in behavior between `git checkout P:\file` and `git checkout P:\sub_dir\file`.  I would of course also like to see `Initialized empty Git repository in P:/.git/` instead of `Initialized empty Git repository in C:/subst_dir/.git/`.

**Comment by hexf00** [comment]
I was asked to use the application in the sandbox. The sandbox seemed to create a virtual drive letter through subst, and the permission to read and write the real path directory was restricted, resulting in the abnormal work of gitlens extension in vscode.

**Comment by dscho** [comment]
There are quite a few code paths in Git that require the real path so that it can be verified whether or not it is inside the worktree or git directory. You can think of this as an invariant. Any change that tries to address this `subst` issue will take pains not to break that invariant again.

**Comment by kcuzner** [comment]
> There are quite a few code paths in Git that require the real path so that it can be verified whether or not it is inside the worktree or git directory. You can think of this as an invariant. Any change that tries to address this subst issue will take pains not to break that invariant again.

Relating to my earlier question:

> I think the linux equivalent of subst would be a bind mount. Does git go so far as to resolve bind mounts to their original paths in linux?

I might be naive here, but considering how most people in this thread complaining (including myself) use bind mounts, it seems like the equivalent is a bind mount. I have never seen git on my Linux install resolve through a bind mount (maybe I'm out of date? I use Arch). Do bind mounts not hit these same cases? And could break them in the same way, if not worse? Bind mounts are far more flexible that `subst`.
